packages/dds/tree/src/feature-libraries/default-schema/defaultFieldKinds.ts

```diff
@@ -171,8 +173,38 @@ export const forbidden = new FieldKindWithEditor(
 	new Set(),
 );
 
+const fieldKindConfigurations: ReadonlyMap<number, FieldKindConfiguration> = new Map([
+	[
+		0,
+		new Map<FieldKindIdentifier, FieldKindConfigurationEntry>([
+			[nodeKey.identifier, { kind: nodeKey, formatVersion: 0 }],
+			[required.identifier, { kind: required, formatVersion: 0 }],
+			[optional.identifier, { kind: optional, formatVersion: 0 }],
+			[sequence.identifier, { kind: sequence, formatVersion: 0 }],
+			[forbidden.identifier, { kind: forbidden, formatVersion: 0 }],
+		]),
+	],
+	[
+		1,
+		new Map<FieldKindIdentifier, FieldKindConfigurationEntry>([
+			[nodeKey.identifier, { kind: nodeKey, formatVersion: 0 }],
+			[required.identifier, { kind: required, formatVersion: 1 }],
+			[optional.identifier, { kind: optional, formatVersion: 1 }],
+			[sequence.identifier, { kind: sequence, formatVersion: 0 }],
+			[forbidden.identifier, { kind: forbidden, formatVersion: 0 }],
+		]),
+	],
+]);
+
+/**
+ * The current configuration for field kinds.
+ * Each field kind has an associated format version that will be used for encoding purposes.
+ */
+export const fieldKindConfiguration: FieldKindConfiguration =
+	fieldKindConfigurations.get(0) ?? fail("Unknown field kind configuration");
```

Comment:
Should we add an entry to TopLevelCodecVersions for this?

Reply:
Perhaps...

See my note at the end of compatibility.md on needing to update O(unversioned depth) places with "new virtual versions" in order to make a change. The analogous flow for something like optional field would be:

- Determine explicit version for EditManager and Message must be rev'd
- Add SharedTreeFormat.v2, add entry in TopLevelCodecVersions mapping to EditManager/Message v2
- Nothing in EditManager's format needs to change, so add entry to its family like this:
![image](https://github.com/microsoft/FluidFramework/assets/5222607/48f5748a-2919-42e1-b853-8666c70a557e)
- Similar case for Message, with both of those going through SharedTreeChangeFamily (this takes a small amount of digging to find without referencing the diagram)
- Bump SharedTreeChangeFamily's codec, which also needs no changes to itself or schemaChangeCodecs:
![image](https://github.com/microsoft/FluidFramework/assets/5222607/db712448-60a7-47ba-a544-8fe11e472cbb)
- Upon examining the codec for modular change family, they'd discover that its version is controlled by `fieldKindConfigurations` and add an entry here.

We can always refactor exactly how different versions resolve as we go in a backwards-compatible way. I'm treating this current task as an exercise in confirming we understand at least one approach to doing that given the timeline for doing this for optional field. If you feel strongly that it would be better to basically centrally manage versions of all codecs (and somehow inject that knowledge to various codecs at creation time) rather than have that policy dictated by calls to resolve different child codecs, we could do that instead.

In my experience, most of the battle in these sorts of refactors is mucking around with layering untangling.

Reply:
No, I don't feel strongly that we would be better off with all implicit versions being centrally managed. It certainly would simplify how one goes about updating versions, but I imagine there's a cost to that design. The good news is that we can change our minds.